ref: add constructor for ECRSpec that takes an ECR image name

etungsten opened the pull request:
*Issue #, if available:* N/A

*Description of changes:*
Adds a constructor for `ECRSpec` that takes an ECR image name of the form `777777777777.dkr.ecr.us-west-2.amazonaws.com/my_image:latest` and constructs an `ECRSpec` struct to be returned.

Adds associated unit tests

*Testing done:*
Unit tests pass.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

samuelkarp commented on ecr/ref.go:
```diff
@@ -70,6 +74,60 @@ func (spec ECRSpec) Registry() string {
 
+// ParseEcrImageNameToRef takes an ECR image name and returns its canonical `ref` format.
+// i.e. ecr.aws/arn:<partition>:ecr:<region>:<account>:repository/<name>:<tag>
+func ParseEcrImageNameToRef(input string) (string, error) {
```
nit: I'd prefer it if this were in a different location in the file.  Either it's a constructor to the `ECRSpec` and should be right below `ParseRef`, or it's not really related to the rest of the functions and should probably be in its own file or at the bottom of this one.

samuelkarp replied:
minor: I think I'd prefer if this became a constructor to `ECRSpec`.  The `ECRSpec` object contains all the information and can be rendered into a ref by its `Canonical` method.

samuelkarp replied:
nit: In Go, [initialisms should be capitalized](https://github.com/golang/go/wiki/CodeReviewComments#initialisms).  Can you capitalize "ECR"?

samuelkarp commented on ecr/ref.go:
```diff
@@ -70,6 +74,60 @@ func (spec ECRSpec) Registry() string {
+	tld := matches[3]
+	region := matches[2]
+	account := matches[1]
+	// If `.cn` TLD, partition should be "aws-cn"
+	// If US gov cloud regions, partition should be "aws-us-gov"
+	// If both of them match, the image source is invalid
+	isCnEndpoint := tld == ".cn"
+	isGovCloudEndpoint := region == "us-gov-west-1" || region == "us-gov-east-1"
+	if isCnEndpoint && isGovCloudEndpoint {
+		return "", err
+	} else if isCnEndpoint {
+		partition = "aws-cn"
+	} else if isGovCloudEndpoint {
+		partition = "aws-us-gov"
+	}
```
I'd prefer that we don't hard-code this logic here.  This is incomplete (it doesn't include the `aws-iso` and `aws-iso-b` partitions) and subject to change as both new regions and new partitions get added.

The AWS SDK for Go keeps [constants for partitions and regions](https://github.com/aws/aws-sdk-go/blob/fa81b5cb4456a1e77f5ccd74d729d1761821aea7/aws/endpoints/defaults.go) and has the [`PartitionForRegion` function](https://github.com/aws/aws-sdk-go/blob/fa81b5cb4456a1e77f5ccd74d729d1761821aea7/aws/endpoints/endpoints.go#L264-L277) that can provide the correct partition given the region.

samuelkarp commented on ecr/ref.go:
```diff
@@ -70,6 +74,60 @@ func (spec ECRSpec) Registry() string {
+	if len(tokens) < 2 {
+		return "", errors.New("No specified name and tag or digest")
+	}
+	fullImageId := tokens[len(tokens)-1]
+	matchDigest, _ := regexp.MatchString(`^[a-zA-Z0-9-_]+@sha256:[A-Fa-f0-9]{64}$`, fullImageId)
+	matchTag, _ := regexp.MatchString(`^[a-zA-Z0-9-_]+:[a-zA-Z0-9\.\-_]{1,128}$`, fullImageId)
```
There's logic for this in the `parseARN` function already (look at how it populates the `object` variable).  I'd recommend refactoring that out and using it in both places.

samuelkarp replied:
It looks like you factored the logic out of `parseARN`, but you're not using it here yet.  Can you use it here too?

etungsten replied:
I use these regex to check for the validity of the tag/digest. The logic in `parseARN` is concerned with separating the repository and the object (tag or digest) and does not check if the tag/digest is valid.

samuelkarp replied:
Why do you want to try and validate them?

samuelkarp replied:
I think this is probably the wrong place to do validation.  You'll get meaningful errors elsewhere.  The digest right now is only `sha256`, but ECR and containerd *could* both gain support for other digests.  The tag length limit is an ECR limit and should be enforced by the ECR service.  Client-side validation like this tightly couples the library to the current implementation details of the things it's integrating with, instead of allowing those components to enforce their own validation.

etungsten replied:
If an image "URI" with a non-existent/invalid tag/digest was passed, the resultant `ECRSpec` will have an empty string for the `object` which would cause any attempts to pull/push image fail. Which I guess is fine? since it'll error out down the line.

etungsten replied:
I'll go ahead and remove these lines.

samuelkarp replied:
It should end up with something in `Object` as long as something was provided after `:` or `@`, but ECR should ultimately be the arbiter of "is it valid".

samuelkarp commented on ecr/ref.go:
```diff
@@ -70,6 +74,60 @@ func (spec ECRSpec) Registry() string {
+// ParseEcrImageNameToRef takes an ECR image name and returns its canonical `ref` format.
+// i.e. ecr.aws/arn:<partition>:ecr:<region>:<account>:repository/<name>:<tag>
+func ParseEcrImageNameToRef(input string) (string, error) {
+	ref := "ecr.aws/"
```
There's a `refPrefix` constant you can use.

samuelkarp commented on ecr/ref.go:
```diff
@@ -35,6 +35,10 @@ const (
+	// Expecting to match ECR image names of the form:
+	// Example 1: 777777777777.dkr.ecr.us-west-2.amazonaws.com/my_image:latest
+	// Example 2: 777777777777.dkr.ecr.cn-north-1.amazonaws.com.cn/my_image:latest
+	ecrRegex = regexp.MustCompile(`(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?.*`)
```
This doesn't currently support the [ECR FIPS endpoints](https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/using-govcloud-endpoints.html), which will have "ecr-fips" in the URL instead of "ecr".

It's fine if you don't want to support that right now, but I'd appreciate at least a TODO comment.

etungsten commented:
Addresses @samuelkarp 's comments.

- Made the conversion function into a constructor for `ECRSpec`
- Leverage `aws-sdk-go` to map regions to the correct partition.
- Refactored the object (tag/digest) parsing into its own function.

Unit tests pass.

samuelkarp commented on ecr/ref.go:
```diff
@@ -104,6 +150,26 @@ func parseARN(a string) (ECRSpec, error) {
+// tag/digest (object) from the resource segment.
+// An example of an ECR ARN resource is "repository/myimage:mytag" and the object returned is "mytag"
+// Another example is "repository/myimage@sha256:47bfdb88c..." and the object returned is "@sha256:47bfdb88c..."
+func parseARNResource(resource *string) string {
```
The argument should be a `string`, not a `*string`.

etungsten replied:
`resource` is being modified in the function; The function strips it of its tag/digest.

samuelkarp replied:
I'd rather you return `(string, string)` than use `*string` as an argument then.  Also might want to call it `splitResource` then since you're splitting the `Resource` and `Object`.

samuelkarp commented on ecr/ref.go:
```diff
@@ -55,6 +61,58 @@ func ParseRef(ref string) (ECRSpec, error) {
+	}
+
+	// Separate out <name>:<tag> for checking validity
+	tokens := strings.Split(input, "/")
```
You probably want this as [`strings.SplitN`](https://golang.org/pkg/strings/#SplitN) since the part after the first `/` should remain together as the resulting `Repository`.  You should then be able to pass this to `parseARNResource` to get the `Object`.

samuelkarp commented on ecr/ref.go:
```diff
@@ -104,6 +150,26 @@ func parseARN(a string) (ECRSpec, error) {
+// tag/digest (object) from the resource segment.
+// An example of an ECR ARN resource is "repository/myimage:mytag" and the object returned is "mytag"
+// Another example is "repository/myimage@sha256:47bfdb88c..." and the object returned is "@sha256:47bfdb88c..."
+func parseARNResource(resource *string) string {
```
This isn't really related to ARNs, so I'd probably call it `parseObject` as that's what it returns for you.

samuelkarp commented on ecr/ref_test.go:
```diff
@@ -179,3 +179,89 @@ func TestImageID(t *testing.T) {
 }
+
+// Test ParseEcrImageNameToRef with a valid ECR image name
+func TestECRImageNameToRefValid(t *testing.T) {
```
nit: the test name should match the function being tested

etungsten commented:
Addresses @samuelkarp 's comments.

- Renamed `NewECRSpec` to `ParseImageURI`
- Removed validation check for tag/digest
- Updated resource parsing function

samuelkarp commented on ecr/ref.go:
```diff
@@ -30,11 +31,17 @@ import (
 const (
 	refPrefix           = "ecr.aws/"
 	repositoryDelimiter = "/"
+	invalidImageURI = "ECRSpec: Invalid image URI"
```
Please start this with "ecrspec:".  See [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings).

samuelkarp commented on ecr/ref.go:
```diff
@@ -104,6 +140,27 @@ func parseARN(a string) (ECRSpec, error) {
+		}
+		retResource = resource[:delimiterIndex[0]]
+	}
+	return object, retResource
```
Can you swap the return order here?  `return resource, object`

samuelkarp commented on ecr/ref.go:
```diff
@@ -104,6 +140,27 @@ func parseARN(a string) (ECRSpec, error) {
+func splitResource(resource string) (string, string) {
+	// remove label & digest
+	var object string
+	retResource := resource
```
You can remove this line and instead mutate the existing `resource` variable.

etungsten commented:
Addresses @samuelkarp 's comments.
